src/chart/rendering/drawlegend.cpp

```diff
@@ -28,48 +32,97 @@ void DrawLegend::draw(Gfx::ICanvas &canvas,
     Gen::ChannelId channelType,
     double weight) const
 {
-	auto contentRect =
+	auto markerWindowRect =
 	    style.contentRect(legendLayout, rootStyle.calculatedSize());
 
-	auto &&info = Info{
-	    .canvas = canvas,
-	    .contentRect = contentRect,
+	auto titleRect =
+	    markerWindowRect.popBottom(style.title.getHeight());
+	auto markerWindowHeight = markerWindowRect.height();
+	auto itemHeight = style.label.getHeight();
+	auto markerSize = style.marker.size->get(itemHeight,
+	    style.label.calculatedSize());
+
+	auto fadeHeight = markerSize;
+	markerWindowRect =
+	    markerWindowRect
+	    + Geom::Rect{{0, -fadeHeight}, {0, 2 * fadeHeight}};
+
+	auto fadeElementPercent = fadeHeight / markerWindowRect.height();
+
+	auto &&info = Info{.canvas = canvas,
+	    .titleRect = titleRect,
+	    .markerWindowRect = markerWindowRect,
+	    .fadeHeight = fadeHeight,
+	    .yOverflow = {},
+	    .yOffset = {},
 	    .type = channelType,
 	    .weight = weight,
-	    .itemHeight = style.label.getHeight(),
-	    .titleHeight = style.title.getHeight(),
-	    .markerSize = style.marker.size->get(contentRect.size.y,
-	        style.label.calculatedSize()),
+	    .itemHeight = itemHeight,
+	    .markerSize = markerSize,
 	    .measure = plot->axises.at(channelType).measure,
 	    .dimension = plot->axises.at(channelType).dimension,
-	};
+	    .colorGradientSetter = {markerWindowRect.leftSide(),
+	        Gfx::ColorGradient{{
+	            {0.0, {}},
+	            {fadeElementPercent / 2.0, {}},
+	            {fadeElementPercent, {}},
+	            {1.0 - fadeElementPercent, {}},
+	            {1.0 - fadeElementPercent / 2.0, {}},
+	            {1.0, {}},
+	        }}}};
+
+	info.yOverflow = markersLegendFullSize(info) - markerWindowHeight;
+	if (std::signbit(info.yOverflow)) info.yOverflow = 0.0;
+	info.yOffset =
+	    style.translateY->get(info.yOverflow, info.itemHeight);
 
 	DrawBackground{{ctx()}}.draw(canvas,
 	    legendLayout,
 	    style,
 	    *events.background,
-	    Events::Targets::legend(channelType));
+	    Events::Targets::legend(channelType, info.yOverflow));
 
 	canvas.save();
-	canvas.setClipRect(contentRect);
 
-	drawTitle(info);
+	canvas.setClipRect(markerWindowRect);
 
 	drawDimension(info);
 
 	drawMeasure(info);
 
 	canvas.restore();
+
+	canvas.save();
+
+	canvas.setClipRect(titleRect);
+
+	drawTitle(info);
+
+	canvas.restore();
+}
+
+void DrawLegend::ColorGradientSetter::operator()(Gfx::ICanvas &canvas,
+    const Geom::AffineTransform &transform,
+    const Gfx::Color &color) const
+{
+	for (auto &stop : modifiableStops) stop.value = color;
+
+	modifiableStops[0].value.alpha = 0.0;
+	modifiableStops[1].value.alpha *= 0.27;
+	modifiableStops[4].value.alpha *= 0.27;
+	modifiableStops[5].value.alpha = 0.0;
+
+	canvas.setBrushGradient(
+	    {transform(line.begin), transform(line.end)},
+	    gradient);
```

Comment:
I think this setter is a bit overcomplicated for the problem. 
What we would like to be able to do is to set a transparency gradient for the label, if needed.  
So why not simply give a `Math::SegmentedFunction<double> transparencyGradient, Geom::Line gradientPosition` parameters for `DrawLabel` and calculate the color gradient in DrawLabel from the calculated text color?

Reply:
because it's not just the label but the marker needs to. And that will be a duplication for the same gradient problem.

Reply:
I see. It's unfortunate. Still could be more clean. I see three problems:
-  This class encapsulates three things (building of the ColorGradient, transforming the line, and setting it to the canvas). I think it would be better if it would have a singular purpose, building the ColorGradient.
- How we build the color gradient is that we have an alpha/transparency gradient, and we get the color from outside.
  the class does not represent this, since it contains a colorgradient, and the transparency gradient is hard coded in the function.
  I think a clearer and more generic approach would be to have a class, which stores an arbitrary alpha gradient and has a method which accepts a color and builds the ColorGradient.
  
Something like this:
  
 ```cpp
 /// legend draw:
 fadeoutLine = markerWindowRect.leftSide()
 fadeoutGradient = AlphaGradient({ 
   {0.0, 0.0}, 
   {fadeElementPercent, 1.0}, 
   {1.0 - fadeElementPercent, 1.0}, 
   {1.0, 0.0} 
 })
 
 // marker draw and label draw
 canvas.setBrushGradient(transform(fadeoutLine), fadeoutGradient.toColorGradient(color))
 ```

Reply:
Also, draw label is a more generic lower level utility class then legend draw. I think it's not a good practice to make the interface of the lower level class more cryptic to optimise code duplication in the higher level user class.